packages/eslint-plugin/lib/rules/no-nested-await.js

```diff
@@ -1,66 +1,70 @@
-/* global module, require */
-/**
- * @author Michael FIG
- * See LICENSE file in root direcotry for full license.
- */
+/* eslint-env node */
 
 'use strict';
 
 /**
- * Imports a specific module.
- *
- * @param {...string} moduleNames - module names to import.
- * @returns {object|null} The imported object, or null.
+ * @author Michael FIG
+ * See LICENSE file in root directory for full license.
  */
-function safeRequire(...moduleNames) {
-  for (const moduleName of moduleNames) {
-    try {
-      // eslint-disable-next-line import/no-dynamic-require,global-require
-      return require(moduleName);
-    } catch (_err) {
-      // Ignore.
-    }
-  }
-  return null;
-}
 
-const CodePathAnalyzer = safeRequire(
-  'eslint/lib/linter/code-path-analysis/code-path-analyzer',
-  'eslint/lib/code-path-analysis/code-path-analyzer',
-);
+const isFunction = node =>
+  [
+    'ArrowFunctionExpression',
+    'FunctionExpression',
+    'FunctionDeclaration',
+  ].includes(node.type);
 
 module.exports = {
   meta: {
     docs: {
-      description:
-        'prevent usage of `await` unless it is a function block statement or destructuring assignment',
+      description: `ensure the first \`await\` in an \`async\` function is non-nested so that it is clear when the synchronous portion of the function is finished`,
       category: 'Possible Errors',
       recommended: true,
       url:
-        'https://github.com/endojs/Jessie/blob/master/packages/eslint-plugin/lib/rules/no-nested-await.js',
+        'https://github.com/endojs/Jessie/blob/main/packages/eslint-plugin/lib/rules/no-nested-await.js',
     },
     type: 'problem',
     fixable: null,
+    messages: {
+      unexpectedNestedAwait:
+        'The first `await` appearing in an async function must not be nested',
+    },
     schema: [],
-    supported: true || CodePathAnalyzer !== null,
+    supported: true,
   },
   create(context) {
-    const isAwaitAllowedInNode = parent => {
-      if (parent.type === 'BlockStatement') {
-        // Find the parent block's node.
-        parent = parent.parent;
+    const foundFirstAwait = new WeakSet();
+
+    const isAwaitAllowedInNode = node => {
```

Comment:
I think this function can be simplified quite a bit (with slight tweaks to the below if we wish to continue rejecting the first `subtlyBalanced` example):
```suggestion
    const isAwaitAllowedInNode = node => {
      // Climb the AST until finding the nearest enclosing function
      // or function-like node.
      let result = true;
      for (; node && !isFunction(node); node = node.parent) {
        // Look only at the first `await` in any given scope.
        if (foundFirstAwait.has(node)) {
          return true;
        }
        foundFirstAwait.add(node);

        // Encountering anything other than an enclosing block
        // is grounds for rejection.
        if (node.type !== 'BlockStatement') {
          result = false;
        }
      }

      if (!node) {
        // Should never reach here because Program is treated like a function
        // for top-level await.
        throw new Error(`unexpected non-Program AST!`);
      }
      return result;
    };
```